driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,134 @@
+                Ok(()) | Err(RecvTimeoutError::Timeout) => {
+                    // Make sure we don't hold the mutex while this blocking call happens.
+                    match price_source.get_prices(&tokens) {
+                        Ok(prices) => price_map_.lock().unwrap().extend(prices),
+                        Err(err) => log::warn!("price_source::get_prices failed: {}", err),
+                    }
+                }
+                // The owning struct has been dropped.
+                Err(RecvTimeoutError::Disconnected) => break,
+            }
+        });
+
+        // Update the prices immediately.
+        sender.send(()).unwrap();
+        (
+            Self {
+                price_map,
+                _channel_to_thread: sender,
+            },
+            join_handle,
+        )
+    }
+}
+
+impl PriceSource for ThreadedPriceSource {
+    /// Non blocking.
+    /// Infallible.
+    fn get_prices(&self, _tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(self.price_map.lock().unwrap().clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::price_source::MockPriceSource;
+    use super::*;
+    use std::sync::atomic;
+
+    const ORDERING: atomic::Ordering = atomic::Ordering::SeqCst;
+
+    lazy_static::lazy_static! {
+        static ref TOKENS: [Token; 3] = [
+            Token::new(0, "0", 0),
+            Token::new(1, "1", 1),
+            Token::new(2, "2", 2),
+        ];
+    }
+
+    /// Drops tps and joins the handle which ensures that the thread exits and
+    /// hasn't panicked. This is important because the mockall expectation
+    /// panics happen in the thread which would otherwise let the tests pass
+    /// when an expectation is violated.
+    fn join(tps: ThreadedPriceSource, handle: JoinHandle<()>) {
+        std::mem::drop(tps);
+        handle.join().unwrap();
+    }
+
+    #[test]
+    fn thread_exits_when_owner_is_dropped() {
+        let mut ps = MockPriceSource::new();
+        ps.expect_get_prices().returning(|_| Ok(HashMap::new()));
+        let (tps, handle) =
+            ThreadedPriceSource::new(TOKENS.to_vec(), ps, Duration::from_secs(std::u64::MAX));
+        // If the thread didn't exit we would wait forever.
+        join(tps, handle);
+    }
+
+    #[test]
+    fn update_triggered_by_interval() {
+        let mut ps = MockPriceSource::new();
+        let start_returning = Arc::new(atomic::AtomicBool::new(false));
+        let start_returning_ = start_returning.clone();
+        ps.expect_get_prices().returning(move |_| {
+            Ok(if start_returning_.load(ORDERING) {
+                hash_map! {TOKENS[0].id => 1}
+            } else {
+                hash_map! {}
+            })
+        });
+
+        let (tps, handle) = ThreadedPriceSource::new(TOKENS.to_vec(), ps, Duration::from_millis(5));
+        assert_eq!(tps.get_prices(&TOKENS[..]).unwrap(), hash_map! {});
+        thread::sleep(Duration::from_millis(10));
```

Comment:
Using sleep statements in unit tests is a smell (it creates brittle tests which pass depending on environment/load).

Generally, I find unit testing threading behaviour hard to do and prefer separating the units so they can each be tested for logic individually (and then we trust that the threading library works as intended).

The way I would test this piece of logic is to move everything inside the thread closure into its own function and invoke this directly in the unit test (thus not testing if channel.recv_timeout works, but we can assume that).

Reply:
The actual logic so simple (`old_prices.extend(new_prices)`) that it's not worth testing.

I agree that sleep statements are not optimal but afaik there is no other way to test multithreaded code. So the question is whether a brittle test is worse than no test. The answer is that I'm not sure. Maybe it would be acceptable to make the tests`#[ignore]`

Reply:
I have thought of a way. Only if we are in the testing configuration, I can add another channel on which the background signals that it has completed its loop. Then in the test we can listen to that channel instead of sleeping. Will implement to see how it looks.

Reply:
> I agree that sleep statements are not optimal but afaik there is no other way to test multithreaded code.

The way I have done it in the past is to separate the threading and the business logic by components or functions and then test the logic in isolation (or abstract the threading component so it can be mocked with a non-threaded executor).

Generally I think a potentially flaky test is worse than no test at all.